flink-examples/src/main/java/io/pravega/examples/flink/alert/HighCountAlerter.java

```diff
@@ -0,0 +1,206 @@
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.windowing.time.Time;
+import org.apache.flink.util.Collector;
+import org.joda.time.DateTime;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+
+/*
+ * This application has the following input parameters
+ *     stream - Pravega stream name to write to
+ *     controller - the Pravega controller URI, e.g., tcp://localhost:9090
+ *                  Note that this parameter is processed in pravega flink connector
+ */
+public class HighCountAlerter {
+
+    // Logger initialization
+    private static final Logger LOG = LoggerFactory.getLogger(HighCountAlerter.class);
+
+    // The application reads data from specified Pravega stream and once every ALERT_INTERVAL (2 seconds)
+    // counts the number of 500 responses in the last ALERT_WINDOW (30 seconds), and generates
+    // alert when the counts exceed ALERT_THRESHOLD (6).
+
+    public static void main(String[] args) throws Exception {
+        LOG.info("Starting HighErrorAlerter...");
+
+        // initialize the parameter utility tool in order to retrieve input parameters
+        ParameterTool params = ParameterTool.fromArgs(args);
+
+        // create Pravega helper utility for Flink using the input paramaters
+        FlinkPravegaParams helper = new FlinkPravegaParams(params);
+
+        // get the Pravega stream from the input parameters
+        StreamId streamId = helper.getStreamFromParam(Constants.STREAM_PARAM,
+                                                      Constants.DEFAULT_STREAM);
+
+        // create the Pravega stream is not exists.
+        helper.createStream(streamId);
+
+        // initialize Flink execution environment
+        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+        // create the Pravega stream reader
+        long startTime = 0;
+        FlinkPravegaReader<String> reader = helper.newReader(streamId, startTime, String.class);
+
+        // add the Pravega reader as the data source
+        DataStream<String> inputStream = env.addSource(reader);
+
+        // create an output sink to stdout for verification
+        //inputStream.print();
+
+        // transform logs
+        DataStream<AccessLog> dataStream = inputStream.map(new ParseLogData());
+
+        // create an output sink to stdout for verification
+        //dataStream.print();
+
+        // get responses and their counts
+        DataStream<ResponseCount> countStream =
+            dataStream.flatMap(new FlatMapFunction<AccessLog, ResponseCount>() {
+                @Override
+                public void flatMap(AccessLog value, Collector<ResponseCount> out) throws Exception {
+                    out.collect(new ResponseCount(value.getStatus(), 1));
```

Comment:
shouldn't we first filter the error log for the response code `500`?

Reply:
Done. i was thinking about doing something processing together with response code, but decide not to make it complicated.

Reply:
We are not making use of the log event time to detect the 500 response code for specific window and it looks like it will be based on the processing time. Should we plan to change the code to make use of the log event time, we could modify something like this?

```
        // initialize the Flink execution environment
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        **env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);**

        // create the Pravega source to read a stream of text
        FlinkPravegaReader<String> reader = FlinkPravegaReader.<String>builder()
                .withPravegaConfig(pravegaConfig)
                .forStream(stream)
                .withDeserializationSchema(PravegaSerialization.deserializationFor(String.class))
                .build();

        // add the Pravega reader as the data source
        DataStream<String> inputStream = env.addSource(reader);

        // transform logs
        DataStream<AccessLog> dataStream =
                inputStream.map(new ParseLogData())
                        .filter(a -> a.getStatus().equals("500")).name("filter");
        DataStream<AccessLog> withTimestampsAndWatermarks =
                dataStream.assignTimestampsAndWatermarks(new EventTimeExtractor(Time.seconds(1000)))
                        .name("watermark");

        DataStream<ResponseCount> countStream =
        withTimestampsAndWatermarks
                .keyBy("verb")
                //.timeWindow(Time.seconds(Constants.ALERT_WINDOW))
                .window(TumblingEventTimeWindows.of(Time.seconds(Constants.ALERT_WINDOW)))
                .fold(new ResponseCount(), new FoldFunction<AccessLog, ResponseCount>() {
                    @Override
                    public ResponseCount fold(ResponseCount accumulator, AccessLog value) throws Exception {
                        accumulator.response = value.getVerb();
                        accumulator.count = accumulator.count + 1;
                        return accumulator;
                    }
                }).name("aggregated-count");

        // create an output sink to stdout for verification
        countStream.print().name("print-count");

        // create alert pattern
        Pattern<ResponseCount,?> pattern500 = Pattern.<ResponseCount>begin("500pattern")
            .where(new SimpleCondition<ResponseCount>() {
                @Override
                public boolean filter(ResponseCount value) throws Exception {
                    return value.count >= Constants.ALERT_THRESHOLD;
                }
            });

        PatternStream<ResponseCount> patternStream = CEP.pattern(countStream, pattern500);

        DataStream<Alert> alertStream = patternStream.select(
            new PatternSelectFunction<ResponseCount, Alert>() {
                @Override
                public Alert select(Map<String, List<ResponseCount>> pattern) throws Exception {
                    ResponseCount count = pattern.get("500pattern").get(0);
                    return new Alert(count.response, count.count, "High 500 responses");
                }
            }).name("alert");

        // create an output sink to stdout for verification
        alertStream.print().name("print-alert");


        // execute within the Flink environment
        env.execute("HighCountAlerter");

    public static class EventTimeExtractor extends BoundedOutOfOrdernessTimestampExtractor<AccessLog> {

        public EventTimeExtractor(Time time) { super(time); }

        @Override
        public long extractTimestamp(AccessLog accessLog) {
            return accessLog.getTimestampMillis();
        }
    }
```

Reply:
thanks for the tip. I plan to come up with a few more examples based on the access log. I will use event time in them later. for this example, i'd like to keep it simple and straightforward to follow.